### DISCUSSION TITLE
GoLang PDK - How to access Config on plugin startup

### DISCUSSION THREAD

**Discussion by borkke** [Kong/kong#8449] [Help] [answered]
I’m working with Go PDK and need to  pass Url parameter to my package.NewCustomLogic but only **once** when the plugin starts. I’m not sure how to access Config object on the startup. Any idea how I could do that? is there some kind of “init” function or could I use main function if yes, how can I access Config object?

```golang
func main() {
	//err := server.StartServer(New, Version, Priority)
	//...
}

type Config struct {
	Url string
}

func New() interface{} {
	return &Config{Url: "..."}
}

// This needs to be instantiated only once on plugin start
var customLogic validator.Validate = *package.NewCustomLogic("<need to pass config.Url here>")

func (conf Config) Access(context *pdk.PDK) {
	//...
}
```

**Comment by fffonion** [comment, marked as answer]
Hi @borkke before answering your question, we can go through a bit on the patterns for a Kong plugin. Each plugin (i.e. your_code.go source) can have multiple "instances" in Kong, the same code can be applied onto different Routes, Services, Consumers etc. When
each phase handler is called, the `config` struct passed in represent the instance of this plugin.
That means you can't have a global `customLogic` that initialized from a config, as that's going to be a static/global variable. A proper way to solve this is to have a "registry" mechaism, it can simply be a `map` with its key being config.Url, then you can
reuse the validator from registry if it's already initialized, or create a new if it's not.

**Reply by borkke** [reply]
Hi, @fffonion thanks for the quick reply. I must admit that I'm not fully with you just yet. Let's try to take this with a more concrete example. 

I want to open up a connection to Redis and I don't want to do this on every HTTP request. I guess this would be wrong then.
```go
func (conf Config) Access(context *pdk.PDK) {
	rdb := redis.NewClient(&redis.Options{
		Addr: config.Url,
		Password: config.Password,
                 ...
	})
}
```

Where you be the best place to open up Redis connection and pass in parameters from the config object?

First I thought it could be in the main method but I can't access config:
```go
var rdb *redis.Client;

func main() {
	//...
	_ := server.StartServer(New, Version, Priority)
	rdb = redis.NewClient(&redis.Options{
		Addr:    ???
		Password: ???
                 ...
	})
}
```

Second, I was thinking there must be some `init` method in the plugin lifecycle that I could use but I'm not finding that.

**Reply by fffonion** [reply]
@borkke There isn't really an `init` method in the plugin lifecycle, Kong plugin instances themselves are usually stateless. The `go-log` plugin (https://github.com/Kong/go-plugins/blob/master/go-log.go) works the same way as yours and can be a useful example.

Basically, you will store those `*redis.Client` in a global map like https://github.com/Kong/go-plugins/blob/master/go-log.go#L48, then you should put mutexes when accessing the redis.Client. The reason behind this is a plugin instance (returned by `New()`) is reused by mutlipled HTTP requests (it matches the nginx worker processes count in Kong).

**Reply by borkke** [reply]
This makes sense, thanks a lot!
